src/windows.rs

```diff
@@ -0,0 +1,480 @@
+                device_name: device.device_name.clone(),
+                function_name: "SetMonitorBrightness",
+                last_error: e.code(),
+            })
+    }
+}
+
+#[derive(Debug)]
+// "Each level is a value from 0 to 100"
+struct IoctlSupportedBrightnessLevels(Vec<u8>);
+
+impl IoctlSupportedBrightnessLevels {
+    fn get_nearest(&self, percentage: u32) -> u8 {
+        self.0
+            .iter()
+            .min_by_key(|&num| (*num as i64 - percentage as i64).abs())
+            .map(|x| *x)
+            .unwrap_or(0)
+    }
+}
+
+fn ioctl_query_supported_brightness(
+    device: &Brightness,
+) -> Result<IoctlSupportedBrightnessLevels, HRESULT> {
+    unsafe {
+        let mut bytes_returned = 0;
+        let mut out_buffer = Vec::<u8>::with_capacity(256);
+        DeviceIoControl(
+            device.file_handle.0,
+            IOCTL_VIDEO_QUERY_SUPPORTED_BRIGHTNESS,
+            std::ptr::null_mut(),
+            0,
+            out_buffer.as_mut_ptr() as *mut c_void,
+            out_buffer.capacity() as u32,
+            &mut bytes_returned,
+            std::ptr::null_mut(),
+        )
+        .ok()
+        .map(|_| {
+            out_buffer.set_len(bytes_returned as usize);
+            IoctlSupportedBrightnessLevels(out_buffer)
+        })
+        .map_err(|e| e.code())
+    }
+}
+
+fn ioctl_query_display_brightness(device: &Brightness) -> Result<u32, SysError> {
+    unsafe {
+        let mut bytes_returned = 0;
+        let mut display_brightness = DISPLAY_BRIGHTNESS::default();
+        DeviceIoControl(
+            device.file_handle.0,
+            IOCTL_VIDEO_QUERY_DISPLAY_BRIGHTNESS,
+            std::ptr::null_mut(),
+            0,
+            (&mut display_brightness as *mut DISPLAY_BRIGHTNESS) as *mut c_void,
+            std::mem::size_of::<DISPLAY_BRIGHTNESS>() as u32,
+            &mut bytes_returned,
+            std::ptr::null_mut(),
+        )
+        .ok()
+        .map(|_| {
+            if display_brightness.ucDisplayPolicy == DISPLAYPOLICY_AC as u8 {
+                // This is a value between 0 and 100.
+                display_brightness.ucACBrightness as u32
+            } else if display_brightness.ucDisplayPolicy == DISPLAYPOLICY_DC as u8 {
+                // This is a value between 0 and 100.
+                display_brightness.ucDCBrightness as u32
+            } else {
+                panic!("Unexpected ucDisplayPolicy");
+            }
+        })
+        .map_err(|e| SysError::GettingBrightnessFailed {
+            device_name: device.device_name.clone(),
+            function_name: "IOCTL_VIDEO_QUERY_DISPLAY_BRIGHTNESS",
+            last_error: e.code(),
+        })
+    }
+}
+
+fn ioctl_set_display_brightness(device: &Brightness, value: u8) -> Result<(), SysError> {
+    // Seems to currently be missing from metadata
+    const DISPLAYPOLICY_BOTH: u8 = 3;
+    unsafe {
+        // TODO: Consider whether this is actually the desired behaviour?
```

Comment:
What is this todo about? Whether to set both AC and DC values?

Reply:
yeah, question was whether the user expects set() to just change the value of whatever the current policy is on or do both.
Possible risk of only setting one is if the power supply changes inbetween the query and set calls, and therefore the functin appears to have no effect, so I think it is best to set both.

Reply:
I agree that setting both is simpler and more robust. Thank you for explaining.